File: thread_performance_test/Java/MergeSort.java
import java.util.*;
import java.time.*;
import java.math.BigDecimal;




public class MergeSort extends Thread{
    static int MAX = 100000;
    static int THREAD_MAX = 4;
    static int initial_time;
    private int final_time;
    static int part = 0;
    
    static int sort_array[];

   


    void merge(int low, int mid, int high) 
{ 
 
   // int *left = malloc(left_length*sizeof(int));
	int left[] = new int[mid - low + 1]; 
    int right[] = new int[high - mid]; 
	// n1 is size of left part and n2 is size 
	// of right part 
	int n1 = mid - low + 1, n2 = high - mid, i, j; 

	// storing values in left part 
	for (i = 0; i < n1; i++) 
		left[i] = sort_array[i + low]; 

	// storing values in right part 
	for (i = 0; i < n2; i++) 
		right[i] = sort_array[i + mid + 1]; 

	int k = low; 
	i = j = 0; 

	// merge left and right in ascending order 
	while (i < n1 && j < n2) { 
		if (left[i] <= right[j]) 
        sort_array[k++] = left[i++]; 
		else
        sort_array[k++] = right[j++]; 
	} 

	// insert remaining values from left 
	while (i < n1) { 
		sort_array[k++] = left[i++]; 
	} 

	// insert remaining values from right 
	while (j < n2) { 
		sort_array[k++] = right[j++]; 
	} 
} 

   public void merge_sort2(int low, int high) 
    { 
        // calculating mid point of array 
        int mid = low + (high - low) / 2; 
        if (low < high) { 
    
            // calling first half 
            merge_sort2(low, mid); 
    
            // calling second half 
            merge_sort2(mid + 1, high); 
    
            // merging the two halves 
            merge(low, mid, high); 
        } 
    } 

    
    
    public void run(){
//System.out.println("part-"+part);
        int thread_part = part++; 

	

	// calculating low and high 
	int low = thread_part * (MAX / 4); 
	int high = (thread_part + 1) * (MAX / 4) - 1; 

	// evaluating mid point 
	int mid = low + (high - low) / 2; 
	if (low < high) { 
		merge_sort2(low, mid); 
		merge_sort2(mid + 1, high); 
		merge(low, mid, high); 
	} 

    }
    public static void main(String [] args){
        int numberOfThreads = 4;
       sort_array = new int[MAX]; 
        
       Random rand = new Random();
        for (int i = 0; i < MAX; i++) 
        sort_array[i] =  rand.nextInt(100) ; 

        System.out.println("Before Array");
        for(int i=0;i<MAX;i++){
            System.out.print(sort_array[i]);
            System.out.print(" ");
        }
        
       long startTime = System.nanoTime();   

        try{
            List<Thread> threads = new ArrayList<>(numberOfThreads);
      
              for (int x = 0; x < numberOfThreads; x++) {
                  Thread t = new Thread(new MergeSort());
                  t.start();
      
      
                 threads.add(t);
                
                    
              // System.out.println("Get Result: " + t.getResult());
              
              }
              for (Thread t : threads) {
                  t.join();
               }
              
          }catch (Exception e) {
              e.printStackTrace();
          }  
      
          MergeSort m = new MergeSort();

       m.merge(0, (MAX / 2 - 1) / 2, MAX / 2 - 1); 
        m.merge(MAX / 2, MAX/2 + (MAX-1-MAX/2)/2, MAX - 1); 
        m.merge(0, (MAX - 1)/2, MAX - 1); 

      long endTime = System.nanoTime();  

      long elapsedTime = endTime - startTime;

      double seconds = (double)elapsedTime / 1000000000;
    
       // initial_time = t1.instant(); 
       System.out.println();
       System.out.println("Sorted Array");
       for(int i=0;i<MAX;i++){
           System.out.print(sort_array[i]);
           System.out.print(" ");
       }
       
       System.out.println("Time Taken: "+BigDecimal.valueOf(seconds).toPlainString());
      
    }


}

File: thread_performance_test/C/merge_sort.c
// CPP Program to implement merge sort using 
// multi-threading 
#include <stdlib.h>
#include <stdio.h>
#include <pthread.h>
#include <unistd.h>
#include <time.h> 

// number of elements in array 
#define MAX 100000

// number of threads 
#define THREAD_MAX 4 


// array of size MAX 
int a[MAX];
//int a[MAX] = {12,11,13,5,6,7}; 
int part = 0; 

// merge function for merging two parts 
void merge(int low, int mid, int high) 
{ 
    int left_length = mid - low + 1;
    int right_length = high - mid;
    int *left = malloc(left_length*sizeof(int));
	//int* left = new int[mid - low + 1]; 
	int* right = malloc(right_length*sizeof(int));

	// n1 is size of left part and n2 is size 
	// of right part 
	int n1 = mid - low + 1, n2 = high - mid, i, j; 

	// storing values in left part 
	for (i = 0; i < n1; i++) 
		left[i] = a[i + low]; 

	// storing values in right part 
	for (i = 0; i < n2; i++) 
		right[i] = a[i + mid + 1]; 

	int k = low; 
	i = j = 0; 

	// merge left and right in ascending order 
	while (i < n1 && j < n2) { 
		if (left[i] <= right[j]) 
			a[k++] = left[i++]; 
		else
			a[k++] = right[j++]; 
	} 

	// insert remaining values from left 
	while (i < n1) { 
		a[k++] = left[i++]; 
	} 

	// insert remaining values from right 
	while (j < n2) { 
		a[k++] = right[j++]; 
	} 
} 

// merge sort function 
void merge_sort2(int low, int high) 
{ 
	// calculating mid point of array 
	int mid = low + (high - low) / 2; 
	if (low < high) { 

		// calling first half 
		merge_sort2(low, mid); 

		// calling second half 
		merge_sort2(mid + 1, high); 

		// merging the two halves 
		merge(low, mid, high); 
	} 
} 

// thread function for multi-threading 
void* merge_sort(void* arg) 
{ 
	// which part out of 4 parts 
	int thread_part = part++; 

	// calculating low and high 
	int low = thread_part * (MAX / 4); 
	int high = (thread_part + 1) * (MAX / 4) - 1; 

	// evaluating mid point 
	int mid = low + (high - low) / 2; 
	if (low < high) { 
		merge_sort2(low, mid); 
		merge_sort2(mid + 1, high); 
		merge(low, mid, high); 
	} 
} 

// Driver Code 
int main() 
{ 
	// generating random values in array 
	for (int i = 0; i < MAX; i++) 
		a[i] = rand() % 100; 


printf("%s\n","Before array:");
	for (int i = 0; i < MAX; i++) {
        printf("%d ",a[i]);
    }
      //  int a[6]  = {12,11,13,5,6,7};

	// t1 and t2 for calculating time for 
	// merge sort 
	clock_t t1, t2; 

	t1 = clock(); 
	pthread_t threads[THREAD_MAX]; 

	// creating 4 threads 
	for (int i = 0; i < THREAD_MAX; i++) 
		pthread_create(&threads[i], NULL, merge_sort, 
										(void*)NULL); 

	// joining all 4 threads 
	for (int i = 0; i < 4; i++) 
		pthread_join(threads[i], NULL); 

	// merging the final 4 parts 
	merge(0, (MAX / 2 - 1) / 2, MAX / 2 - 1); 
	merge(MAX / 2, MAX/2 + (MAX-1-MAX/2)/2, MAX - 1); 
	merge(0, (MAX - 1)/2, MAX - 1); 

	t2 = clock(); 

printf("%s\n");
	// displaying sorted array 

    printf("%s\n","Sordted array:");
	for (int i = 0; i < MAX; i++) {
        printf("%d ",a[i]);
    }
	// time taken by merge sort in seconds 
    printf("Time Taken: %f\n",(t2 - t1) / 
			(double)CLOCKS_PER_SEC);

	return 0; 
}
